### PR TITLE
 Add a job to bump the jenkins rpm version and build it

### PR DESCRIPTION
**Pull request by smunilla**

### CODE REVIEW

**Comment by smunilla** [comment]
@jupierce Should be ready for eyes now.

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                    description: 'OCP target release',
+                    $class: 'hudson.model.ChoiceParameterDefinition',
+                    choices: ['rhaos-3.7-rhel-7', 'rhaos-3.6-rhel-7'].join('\n'),
+                    defaultValue: 'rhaos-3.7-rhel-7'
+                ],
+                [
+                    name: 'MAIL_LIST_SUCCESS',
+                    description: 'Success Mailing List',
+                    $class: 'hudson.model.StringParameterDefinition',
+                    defaultValue: 'jupierce@redhat.com,smunilla@redhat.com,ahaile@redhat.com,gmontero@redhat.com,bparees@redhat.com,pep@redhat.com'
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: no more pep :(

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env groovy
+
+properties(
+    [
+        [ $class : 'ParametersDefinitionProperty', parameterDefinitions: [
+                [
+                    name: 'JENKINS_VERSION',
+                    description: 'Minimum required Jenkins version',
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: Seems like the wrong description. I also like to put an example valid value in the description.

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                    $class: 'hudson.model.StringParameterDefinition',
+                ],
+                [
+                    name: 'OCP_RELEASE',
+                    description: 'OCP target release',
+                    $class: 'hudson.model.ChoiceParameterDefinition',
+                    choices: ['rhaos-3.7-rhel-7', 'rhaos-3.6-rhel-7'].join('\n'),
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: Might as well fill this out through 3.9

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                    name: 'MAIL_LIST_SUCCESS',
+                    description: 'Success Mailing List',
+                    $class: 'hudson.model.StringParameterDefinition',
+                    defaultValue: 'jupierce@redhat.com,smunilla@redhat.com,ahaile@redhat.com,gmontero@redhat.com,bparees@redhat.com,pep@redhat.com'
+                ],
+                [
+                    name: 'MAIL_LIST_FAILURE',
+                    description: 'Failure Mailing List',
+                    $class: 'hudson.model.StringParameterDefinition',
+                    defaultValue: 'jupierce@redhat.com,smunilla@redhat.com,ahaile@redhat.com,gmontero@redhat.com,bparees@redhat.com,pep@redhat.com'
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: pep again

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    mail(
+        to: "${MAIL_LIST_SUCCESS}",
+        from: "aos-cd@redhat.com",
+        replyTo: 'jupierce@redhat.com',
+        subject: "jenkins plugins RPM for OCP ${OCP_RELEASE} updated in dist-git",
+        body: """The Jenkins plugins RPM for OCP ${OCP_RELEASE} has been updated in dist-git:
+${distgit_link}
+
+Minimum Jenkins version: ${JENKINS_VERSION}
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: Minimum?

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        ],
+        disableConcurrentBuilds()
+    ]
+)
+
+def mail_success() {
+
+    jenkins_major = JENKINS_VERSION.tokenize('.')[0].toString()
+
+    distgit_link = "http://pkgs.devel.redhat.com/cgit/rpms/jenkins-${jenkins_major}-plugins/?h=rhaos-${OCP_RELEASE}-rhel-7"
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: plugins?

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        to: "${MAIL_LIST_SUCCESS}",
+        from: "aos-cd@redhat.com",
+        replyTo: 'jupierce@redhat.com',
+        subject: "jenkins plugins RPM for OCP ${OCP_RELEASE} updated in dist-git",
+        body: """The Jenkins plugins RPM for OCP ${OCP_RELEASE} has been updated in dist-git:
+${distgit_link}
+
+Minimum Jenkins version: ${JENKINS_VERSION}
+
+Plugin list:
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: plugins?

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+def mail_failure(err) {
+
+    mail(
+        to: "${MAIL_LIST_FAILURE}",
+        from: "aos-cd@redhat.com",
+        replyTo: 'jupierce@redhat.com',
+        subject: "Error during jenkins plugin RPM update on dist-git",
+        body: """The job to update the jenkins plugins RPM in dist-git encountered an error:
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: plugins?

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    buildlib.kinit()       // Sets up credentials for dist-git access
+
+    try {
+        stage ("prepare workspace") {
+
+            // Temporary file to write the plugin list to
+            tmpdir = pwd tmp:true
+            scripts_dir = "${env.WORKSPACE}/hacks/bump-jenkins"
+
+            writeFile file: plugin_file, text: PLUGIN_LIST
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: plugins?

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+        stage ("collect plugins") {
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: stage names should be relevant to operation.

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+            withEnv(["PATH+SCRIPTS=${scripts_dir}"]) {
+                sh "bump-jenkins.sh ${JENKINS_VERSION} ${OCP_RELEASE}"
+            }
+        }
+
+        stage ("update dist-git") {
+            withEnv(["PATH+SCRIPTS=${scripts_dir}"]) {
+                sh "update-dist-git.sh ${JENKINS_VERSION} ${OCP_RELEASE} ${plugin_dir}"
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: I don't think this stage exists for your job.

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+            }
+        }
+
+        stage ("cleanup") {
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: also not needed.

File: jobs/devex/jenkins-bump-version/README.md [hunk 13]
```diff
@@ -0,0 +1,13 @@
+This job invokes the [scripts to update the jenkins base RPM](../../hacks/update-jenkins-plugins/README.MD).
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/README.md]: Needs to be re-written.

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 14]
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: Make sure this script is executable according to git. Hard to tell with github.

File: jobs/devex/jenkins-bump-version/Jenkinsfile [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+    def buildlib = load( "pipeline-scripts/buildlib.groovy" )
+    buildlib.kinit()       // Sets up credentials for dist-git access
+
+    try {
+        stage ("prepare workspace") {
+
+            // Temporary file to write the plugin list to
+            tmpdir = pwd tmp:true
+            scripts_dir = "${env.WORKSPACE}/hacks/bump-jenkins"
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/Jenkinsfile]: Looks like actual directory is `bump-jenkins-version`

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+# rhpkg build
+build_jenkins() {
+  rhpkg build
+}
+
+# Make sure they passed something in for us
+if [ "$#" -lt 2 ] ; then
+  usage
+fi
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: exit 1

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    echo "Usage `basename $0` <jenkins-war-version> <branch>" >&2
+    echo >&2
+    echo " Example: ./bump-jenkins.sh 2.73.1 rhaos-3.7-rhel-7" >&2
+    exit 1
+}
+
+# clone jenkins distgit
+# switch to latest branch 
+setup_dist_git() {
+  container="jenkins"
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: `container` isn't accurate. And can't this use $JENKINS_DIST_GIT ?

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+  workingdir=$(dirname $(realpath $0))/working
+
+  if ! klist &>${workingdir}/logs/${container}.output ; then
+    echo "Error: Kerberos token not found." ; popd &>${workingdir}/logs/${container}.output ; exit 1
+  fi
+
+  cd ${workingdir}
+  rhpkg ${USER_USERNAME} clone "${container}" &>${workingdir}/logs/${container}.output
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: won't you need to cd into this directory for some of your other rhpkg invocations to work?

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 19]
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
+
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: Add `set -euo pipefail`

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    echo >&2
+    echo " Example: ./bump-jenkins.sh 2.73.1 rhaos-3.7-rhel-7" >&2
+    exit 1
+}
+
+# clone jenkins distgit
+# switch to latest branch 
+setup_dist_git() {
+  container="jenkins"
+  workingdir=$(dirname $(realpath $0))/working
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]:
```
rm -rf $workingdir
mkdir -p $workingdir/logs
```

File: hacks/bump-jenkins-version/bump-jenkins.sh (hunk 19 above)
Comment (smunilla) [reply hacks/bump-jenkins-version/bump-jenkins.sh]: ✔️

File: hacks/bump-jenkins-version/bump-jenkins.sh (hunk 18 above)
Comment (smunilla) [reply hacks/bump-jenkins-version/bump-jenkins.sh]: ✔️

File: hacks/bump-jenkins-version/bump-jenkins.sh (hunk 16 above)
Comment (smunilla) [reply hacks/bump-jenkins-version/bump-jenkins.sh]: ✔️

File: hacks/bump-jenkins-version/bump-jenkins.sh (hunk 17 above)
Comment (smunilla) [reply hacks/bump-jenkins-version/bump-jenkins.sh]: ✔️

File: hacks/bump-jenkins-version/bump-jenkins.sh (hunk 20 above)
Comment (smunilla) [reply hacks/bump-jenkins-version/bump-jenkins.sh]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 2 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 3 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 1 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 4 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 6 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 5 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 7 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 8 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 15 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 9 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 10 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 11 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/Jenkinsfile (hunk 12 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/Jenkinsfile]: ✔️

File: jobs/devex/jenkins-bump-version/README.md (hunk 13 above)
Comment (smunilla) [reply jobs/devex/jenkins-bump-version/README.md]: ✔️

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+
+  cd ${workingdir}
+  rhpkg ${USER_USERNAME} clone "${JENKINS_DIST_GIT}" &>${workingdir}/logs/${JENKINS_DIST_GIT}.output
+  if [ -d ${JENKINS_DIST_GIT} ] ; then
+    cd ${JENKINS_DIST_GIT}
+    pushd ${JENKINS_DIST_GIT} >${workingdir}/logs/${JENKINS_DIST_GIT}.output
+    rhpkg switch-branch "${BRANCH}" &>${workingdir}/logs/${JENKINS_DIST_GIT}.output
+    popd >${workingdir}/logs/${JENKINS_DIST_GIT}.output
+  else
+    echo " Failed to clone package: ${JENKINS_DIST_GIT}"
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: exit 1

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+    echo "Usage `basename $0` <jenkins-war-version> <branch>" >&2
+    echo >&2
+    echo " Example: ./bump-jenkins.sh 2.73.1 rhaos-3.7-rhel-7" >&2
+    exit 1
+}
+
+# clone jenkins distgit
+# switch to latest branch 
+setup_dist_git() {
+  workingdir=$(dirname $(realpath $0))/working
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: does $0 work in this context?

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+  rm -rf $workingdir
+  mkdir -p $workingdir/logs
+
+
+  if ! klist &>${workingdir}/logs/${JENKINS_DIST_GIT}.output ; then
+    echo "Error: Kerberos token not found." ; popd &>${workingdir}/logs/${JENKINS_DIST_GIT}.output ; exit 1
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: popd?

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+  fi
+
+  cd ${workingdir}
+  rhpkg ${USER_USERNAME} clone "${JENKINS_DIST_GIT}" &>${workingdir}/logs/${JENKINS_DIST_GIT}.output
+  if [ -d ${JENKINS_DIST_GIT} ] ; then
+    cd ${JENKINS_DIST_GIT}
+    pushd ${JENKINS_DIST_GIT} >${workingdir}/logs/${JENKINS_DIST_GIT}.output
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: pushd is wrong now.

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+    rhpkg switch-branch "${BRANCH}" &>${workingdir}/logs/${JENKINS_DIST_GIT}.output
+    popd >${workingdir}/logs/${JENKINS_DIST_GIT}.output
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: popd not needed

File: hacks/bump-jenkins-version/bump-jenkins.sh [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+  rm -rf $workingdir
+  mkdir -p $workingdir/logs
+
+
+  if ! klist &>${workingdir}/logs/${JENKINS_DIST_GIT}.output ; then
+    echo "Error: Kerberos token not found." ; popd &>${workingdir}/logs/${JENKINS_DIST_GIT}.output ; exit 1
+  fi
+
+  cd ${workingdir}
+  rhpkg ${USER_USERNAME} clone "${JENKINS_DIST_GIT}" &>${workingdir}/logs/${JENKINS_DIST_GIT}.output
```
Comment (jupierce) [inline hacks/bump-jenkins-version/bump-jenkins.sh]: I think all rhpkg invocations are going to need `--user ocp-build`

File: jobs/devex/jenkins-bump-version/README.md [hunk 27]
```diff
@@ -0,0 +1,13 @@
+This job invokes the [script to update the jenkins base RPM](../../hacks/update-jenkins/README.MD).
+
+It needs the following parameters:
+
+- Jenkins version. Version to upgrade to
+
+- OCP branch. Used to determine the target dist-git branch
+
+- plugin list, one plugin per line, in the form pluginname:version (see scripts for more details)
```
Comment (jupierce) [inline jobs/devex/jenkins-bump-version/README.md]: plugin

**Comment by jupierce** [comment]
Fixes made in another pull.

**Pull request merged**